### PR TITLE
Throw on errors in replay tool instead of logging to console

### LINKED ISSUE #3107: Snapshot tests are not always failing from errors
Errors encountered in the threads are not always propagating and causing the tests to fail.

The error I consistently repro (every time) was an op handling error (attach message was throwing because of some bad code). The output would include the error details, because the logger was printing them, but the test itself wasn't failing.

### PR DESCRIPTION
**Pull request by heliocliu**
Fixes #3107

Currently the telemetry logger used in the replay tool will print error details (such as from a container error) as console errors.  The snapshot tests do not have any specific assertions and instead rely on promise resolution to pass/fail, so this functionally hides swallows failures and causes failing tests to report succeeded.  Change the logger to re-throw the error in these cases, causing tests to fail as expected.

Potentially all console errors should be made to fail tests, but that's a different discussion.

### CODE REVIEW

**Comment by vladsud** [comment]
This might be right change, but I do not have confidence that exception will not be eaten somewhere and ignored.
Also I remember there were issues with the tool around reporting unhandled exceptions and actually making node process return non-zero error code in such cases.
I guess it would be great to confirm that it indeed improves the problem space.
And maybe consider defense in depth - i.e. log that error happened, and when app run is over, fail it if we recorded that there was a failure encountered earlier.

File: packages/tools/replay-tool/src/replayMessages.ts [hunk 1]
```diff
@@ -125,11 +125,10 @@ class Logger implements ITelemetryBaseLogger {
...
             delete event.stack;
-            console.error(`An error has been logged from ${this.containerDescription}!`);
-            console.error(event);
-            if (stack) {
-                console.error(stack);
-            }
+            const err = `An error has been logged from ${this.containerDescription}!\n
+                        ${JSON.stringify(event)}\nStack:\n${stack}`;
+            // throw instead of printing an error to fail tests
+            throw new Error(err);
```
Comment (vladsud) [inline packages/tools/replay-tool/src/replayMessages.ts:131]: I think you can do something like that:
```
error = new Erorr("...");
error.stack = stack;
throw error;
```
maybe casting to any is required, not sure

**Review by vladsud — APPROVED** [review]

**Review by arinwt — APPROVED** [review]

**Comment by vladsud** [comment]
BTW, the issue #3107 does not talk about that code, as far as I understand. 
It talks about this code, I believe:  https://github.com/microsoft/FluidFramework/blob/main/packages/test/snapshots/src/replayMultipleFiles.ts#L165

And I think @jatgarg was poking and addressing some of the problems related to multi-threading and error propagation.

File: packages/tools/replay-tool/src/replayMessages.ts (hunk 1 above)
Comment (arinwt) [reply packages/tools/replay-tool/src/replayMessages.ts:131]: I thought I had seen code like that somewhere, but I can't find it now.

**Comment by heliocliu** [comment]
> 
> 
> BTW, the issue #3107 does not talk about that code, as far as I understand.
> It talks about this code, I believe: https://github.com/microsoft/FluidFramework/blob/main/packages/test/snapshots/src/replayMultipleFiles.ts#L165
> 
> And I think @jatgarg was poking and addressing some of the problems related to multi-threading and error propagation.

@arinwt suggested undoing his linked commit to repro the errors running snapshot tests, and the ones I ended up with were the ones getting swallowed in replayMessages.  `replayMultipleFiles.ts` looks fine because any promise rejection isn't being caught it looks like, and should propagate up into the test method.  So this fixes the described issue around snapshot tests, but it may not fix all cases of errors getting swallowed (probably just iterate in this case as we find them), and could potentially disrupt replay-tool usage if container errors are expected.  Do we know if that's the case (that container errors are expected and should not block replay-tool from continuing to run)?  This would need to be more flexible if so, such as by returning errors at the end like you described

**Comment by vladsud** [comment]
Hard to say. I'd think that if tests are passing, and we can take some recent Fluid file (downloaded through fetch tool) and run through it with replay tool - as long as that's successful, we should be good.
That said, Mark & I tried to run it today, and it failed (unrelated to your changes), so validation there is not easily possible today. It failed for two reasons:
1. It was not updated for detached workflow - it assumes starting with NullRuntime is the way to go.
2. New components were added that it is not aware about - it failed on LastEdited factory being missing.

I've opened issues to track them:
https://github.com/microsoft/FluidFramework/issues/3738
https://github.com/microsoft/FluidFramework/issues/3739

**Comment by jatgarg** [comment]
> BTW, the issue #3107 does not talk about that code, as far as I understand.
> It talks about this code, I believe: https://github.com/microsoft/FluidFramework/blob/main/packages/test/snapshots/src/replayMultipleFiles.ts#L165
> 
> And I think @jatgarg was poking and addressing some of the problems related to multi-threading and error propagation.

Few months back when I enabled the snapshot tests, I saw that error in there were not causing the tests to fail.
https://github.com/microsoft/FluidFramework/pull/2214/files
This was my PR which then fixed that issue.

**Comment by heliocliu** [comment]
checking this in as is; if it causes issues with replay-tool once it's working again i can revisit/revert as needed.

**Pull request merged**